_posts/2020-01-23-scatter-gather-pt2.md

```diff
@@ -0,0 +1,142 @@
+
+The next step is to wait for all responses to arrive.
+
+### Counting responses
+
+This step can be implemented in many different ways. In this case, the saga will trigger a timeout to check the number of arrived responses. This check repeats until the expected number of responses are stored in the database, at which point we can aggregate the data and finish the gather process:
+
+```
+class ScatterGatherSaga : Saga<ScatterGatherSagaData>, 
+    IAmStartedByMessages<StartSagaCommand>,
+    IHandleTimeouts<CheckStatus>
+{
+    public async Task Handle(StartSagaCommand message, IMessageHandlerContext context)
+    {
+        Data.NumberOfMessages = message.NumberOfMessages;
+
+        for (int i = 0; i < message.NumberOfMessages; i++)
+        {
+            await context.SendLocal(new RequestResponseCommand()
+            {
+                BatchId = message.Id,
+                RequestId = i
+            });
+        }
+
+        // set the first timeout
+        await RequestTimeout<CheckStatus>(context, TimeSpan.FromSeconds(5));
+    }
+
+    public async Task Timeout(CheckStatus state, IMessageHandlerContext context)
+    {
+        var mongoSession = context.SynchronizedStorageSession.GetClientSession();
+        var mongoCollection = mongoSession.Client.GetDatabase(Program.DatabaseName)
+            .GetCollection<ResponseResult>("results");
+
+        var totalDocuments = await mongoCollection.CountDocumentsAsync(mongoSession,
+            Builders<ResponseResult>.Filter.Eq(d => d.SagaId, Data.Identifier));
+        if (totalDocuments >= Data.NumberOfMessages)
+        {
+            // all responses received, aggregate data
+            var responses = await mongoCollection
+                .FindAsync<ResponseResult>(mongoSession, 
+                    Builders<ResponseResult>.Filter.Eq(r => r.SagaId, Data.Identifier));
+
+            int result = responses.ToEnumerable().Sum(response => response.Result);
+            Console.WriteLine("Total result is: " + result);
+            MarkAsComplete();
+        }
+        else
+        {
+            // request another timeout until we have all the response messages
+            await RequestTimeout<CheckStatus>(context, TimeSpan.FromSeconds(1));
+        }
+    }
+
+    // mapping omitted
+}
+```
+
+## Summary
+
+In this blog post we've looked at a more complex approach to deal with scatter-gather scenarios in NServiceBus by extracting the gather process from the saga. With this approach we can eliminate the majority of problems caused by pessimistic locking or optimistic concurrency control and improve the performance or even avoid expensive retries and error messages. The outlined approach is supposed to give some ideas on how to avoid the existing limitations sagas in regards to concurrency, it is not a complete example nor does it cover all of the possible scenarios.
```

Comment:
```suggestion
In this blog post, we've looked at a more complex approach to deal with scatter-gather scenarios in NServiceBus by extracting the gathering process from the saga into a dedicated handler. With this approach, we can eliminate most of the problems caused by pessimistic locking or optimistic concurrency control and improve the performance or even avoid expensive retries and failed messages. 

The outlined approach is supposed to give some ideas on how to avoid the existing limitations of sagas in regards to concurrency, it is not a complete example nor does it cover all of the possible scenarios.
```